### PR TITLE
Bump to latest binutils==2.43

### PR DESCRIPTION
**Pull request by CTM1**
Changes:

- Bumped to 2.43, which required some modifications:
  - At one point, libbfd's `init_disassemble_info` started taking 4 arguments instead of 3. The 4th argument is a styled fprintf function, added a printf wrapper that ignores the style in the same fashion as (https://git.kernel.org/pub/scm/linux/kernel/git/bpf/bpf-next.git/commit/?id=600b7b26c07a070d0153daa76b3806c1e52c9e00).
  - Certain flag incompatibilities with later GCC versions (>10.something) required the `-fcommon` flag to avoid multiple definition errors. Added a `execute_with_env` function to the build script to accommodate this (https://github.com/easybuilders/easybuild-easyconfigs/issues/11988).
  - Updated the C examples. The documentation is outdated but I haven't had the time to update it. Had to modify the Makefile slightly to make them work
  - Linking `zstd` and `sframe` native libs on the build script.
  - Added checks to the C helpers and set default flavor to intel, this doesn't work for unsupported architectures like `mep` which still use AT&T syntax (who cares about `mep` anyway).
  - Valgrind warned about some uninitialized values in the `configure_disassemble_info_buffer` which caused some heisenbugs (worked when I printf the values to check them but SIGSEVd when I didn't)  so I modified the FFI and the helper function to take extra arguments and to properly initialize those values with zeroes. Same for the allocation helpers.
  - Small style changes suggested by the compiler to avoid warnings.
  
  I'm using this crate for a personal project and these modifications will do for now, I might add some more bindings in the future.
  
  All tests are passing, be my guest to check the binutils hash used in the build script.

### CODE REVIEW

**Comment by CTM1** [comment]
Hello, is there anything missing you would like to see ?

**Comment by guedou** [comment]
I should have answered earlier but I am super busy right now =\

I will need a bit more time to review your PR, but it looks great on my first overview.

I am really happy to read that you are using this crate!

File: resources/README.md [hunk 1]
```diff
@@ -10,24 +10,15 @@ opcodes libraries from C:
 3. [test_buffer_mep](examples/src/test_buffer_mep.c): similar to the second
    example with a more exotic architecture and binutils builtins
 
+**THIS IS OUTDATED**
```
Comment (guedou) [inline resources/README.md]:
```suggestion
**As of September 2024, the following documentation is outdated:**
```

File: resources/examples/Makefile [hunk 2]
```diff
@@ -1,17 +1,15 @@
-# Copyright (C) 2018 Guillaume Valadon <guillaume@valadon.net>
```
Comment (guedou) [inline resources/examples/Makefile:1]: Can you keep this line? You can add you name into `Cargo.toml`

File: src/opcodes.rs [hunk 3]
```diff
@@ -98,29 +98,38 @@ impl DisassembleInfo {
         offset: u64,
     ) -> Result<(), Error> {
         utils::check_null_pointer(self.info, "info pointer is null!")?;
-
+
```
Comment (guedou) [inline src/opcodes.rs]: Can you revert this change?

File: src/opcodes.rs [hunk 4]
```diff
@@ -98,29 +98,38 @@ impl DisassembleInfo {
         offset: u64,
     ) -> Result<(), Error> {
         utils::check_null_pointer(self.info, "info pointer is null!")?;
-
+    
         unsafe {
             let ptr = buffer.as_ptr();
             utils::check_null_pointer(ptr, "buffer pointer is null!")?;
-
+
```
Comment (guedou) [inline src/opcodes.rs]: Can you revert this change?

File: src/opcodes.rs [hunk 5]
```diff
@@ -98,29 +98,38 @@ impl DisassembleInfo {
...
+                arch,
+                mach,
+                offset,
+                len as u64,
+                ptr,
+            );
+    
+            if section.is_null() {
                 return Err(Error::DisassembleInfoError(
                     "set_buffer() malloc error!".to_string(),
```
Comment (guedou) [inline src/opcodes.rs]: I
```suggestion
                    "helpers::configure_disassemble_info_buffer() malloc error!".to_string(),
```

**Review by guedou — CHANGES_REQUESTED** [review]
This looks good to me.  I made some minor comments.

**Review by guedou — APPROVED** [review]

**Comment by guedou** [comment]
I had to apply the following patch to make it compile. Which rustc, architecture and OS are you using?

```
git diff src/bfd.rs
diff --git a/src/bfd.rs b/src/bfd.rs
index e2824f4..ee334d0 100644
--- a/src/bfd.rs
+++ b/src/bfd.rs
@@ -180,7 +180,7 @@ pub fn arch_list() -> Vec<String> {
                 stop = true;
                 break;
             }
-            let arch = unsafe { CStr::from_ptr(*item as *const i8).to_str() };
+            let arch = unsafe { CStr::from_ptr(*item as *const u8).to_str() };
             match arch {
                 Ok(s) => ret_vec.push(s.to_string()),
                 Err(_) => ret_vec.push("arch_list() - from_ptr() error !".to_string()),
```

**Comment by CTM1** [comment]
I've ran `cargo build` successfully in two different machines, both Linux, one ubuntu-based and another arch-based.

```
Pop!_OS 22.04 LTS x86_64

$ cargo -V
cargo 1.79.0 (ffa9cf99a 2024-06-03)                                               
$ rustc -V
rustc 1.79.0 (129f3b996 2024-06-10)
$ echo -n $TARGETS
$ %
```

```
EndeavourOS x86_64

$ cargo -V
cargo 1.80.1 (376290515 2024-07-16)
$ rustc -V
rustc 1.80.1 (3f5fd8dd4 2024-08-06)
$ echo -n $TARGETS
$ %
```

**Comment by CTM1** [comment]
I suppose on some systems, char in C is treated as signed while on others, it's unsigned, must be careful with this from now on...

Latest commit should fix the issue, lmk

Edit:

Doc specifies i8...
https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_ptr

Source code signature wants c_char...
https://doc.rust-lang.org/1.81.0/src/core/ffi/c_str.rs.html#282

These should be the same

https://docs.rs/libc/0.2.23/libc/type.c_char.html

https://stackoverflow.com/questions/44436515/should-i-use-libcc-char-or-stdosrawc-char

**Review by guedou — COMMENTED** [review]
This looks good to me.

**Comment by guedou** [comment]
This looks good to me but I still have an issue with the `build.rs` file on my aarch64 Ubuntu system. I will provide a fix.

Do you also have warnings? If yes, can you fix them.
```
$ cargo run --example test_buffer
warning: no edition set: defaulting to the 2015 edition while the latest is 2021
warning: binutils@0.1.1: src/helpers.c: In function ‘configure_disassemble_info’:
warning: binutils@0.1.1: src/helpers.c:134:70: warning: cast between incompatible function types from ‘void (*)(void *, enum disassembler_style,  const char *, ...)’ to ‘int (*)(void *, enum disassembler_style,  const char *, ...)’ [-Wcast-function-type]
warning: binutils@0.1.1:   134 |     init_disassemble_info(info, stdout, (fprintf_ftype) copy_buffer, (fprintf_styled_ftype) copy_buffer_styled);
warning: binutils@0.1.1:       |                                                                      ^
warning: binutils@0.1.1: src/helpers.c: In function ‘configure_disassemble_info_buffer’:
warning: binutils@0.1.1: src/helpers.c:161:70: warning: cast between incompatible function types from ‘void (*)(void *, enum disassembler_style,  const char *, ...)’ to ‘int (*)(void *, enum disassembler_style,  const char *, ...)’ [-Wcast-function-type]
warning: binutils@0.1.1:   161 |     init_disassemble_info(info, stdout, (fprintf_ftype) copy_buffer, (fprintf_styled_ftype) copy_buffer_styled);
warning: binutils@0.1.1:       |                                                                      ^
warning: binutils@0.1.1: src/helpers.c: In function ‘copy_buffer’:
warning: binutils@0.1.1: src/helpers.c:68:1: warning: control reaches end of non-void function [-Wreturn-type]
warning: binutils@0.1.1:    68 | }
warning: binutils@0.1.1:       | ^
```

**Comment by CTM1** [comment]
I don't have the warnings you've mentioned, could you give me your `rustc` version or any extra flags you're using in build so I can replicate ?
I've tried to provide a fix regardless, tell me if it works for you

**Review by guedou — APPROVED** [review]

File: build.rs [hunk 6]
```diff
@@ -185,8 +210,11 @@ fn main() {
     println!("cargo:rustc-link-lib=static=opcodes");
     println!("cargo:rustc-link-lib=static=iberty");
 
-    // Link to zlib
+    // Link to zlib, zstd and sframe
     println!("cargo:rustc-link-search=native=/usr/lib/"); // Arch Linux
     println!("cargo:rustc-link-search=native=/usr/lib/x86_64-linux-gnu/"); // Debian based
```
Comment (guedou) [inline build.rs]:
```suggestion
    println!("cargo:rustc-link-search=native=/usr/lib/{}-linux-gnu/", // Debian based
             env::var("CARGO_CFG_TARGET_ARCH").unwrap());
```

File: src/helpers.c [hunk 7]
```diff
@@ -21,22 +21,101 @@ void buffer_to_rust(char *buffer);
...
+The 4th argument is a styled fprintf function.
+
+Here we will implement copy_buffer_styled, which will just be a wrapper that ignores style changes.
+***/
+int copy_buffer(void *useless, const char *format, ...);
+int copy_buffer_styled(void* user_data, enum disassembler_style style, const char* format, ...);
+
+int copy_buffer(void* useless, const char* format, ...) {
+    // Use the parameter to prevent optimization
+    (void)useless;  // mark as unused
```
Comment (guedou) [inline src/helpers.c]:
```suggestion
    (void) useless;  // mark as unused
```

File: src/helpers.c [hunk 8]
```diff
@@ -21,22 +21,101 @@ void buffer_to_rust(char *buffer);
...
+    
+    // buffer_to_rust uses to_string() which is now owned (cloned) so we can safely free here.
+    free(buffer_asm);
+
+    return result;
+}
+
+int copy_buffer_styled(void* user_data, enum disassembler_style style, const char* format, ...) {
+    // Use the parameter to prevent optimization
+    (void)user_data;  // mark as unused
```
Comment (guedou) [inline src/helpers.c]:
```suggestion
    (void) user_data;  // mark as unused
```

File: src/helpers.c [hunk 9]
```diff
@@ -21,22 +21,101 @@ void buffer_to_rust(char *buffer);
...
+    (void)style;
```
Comment (guedou) [inline src/helpers.c:79]:
```suggestion
    (void) style;
```

File: src/helpers.c [hunk 10]
```diff
@@ -49,53 +128,77 @@ void show_buffer(struct disassemble_info *info) {
...
 bfd_boolean configure_disassemble_info(struct disassemble_info *info, asection *section, bfd *bfdFile) {
+    if (info == NULL || section == NULL || bfdFile == NULL) {
+        fprintf(stderr, "Error: Null pointer passed to configure_disassemble_info\n");
+        return FALSE;
+    }
+
     /* Construct and configure the disassembler_info class using stdout */
-  
-    init_disassemble_info (info, stdout, (fprintf_ftype) copy_buffer);
+    init_disassemble_info(info, stdout, (fprintf_ftype) copy_buffer, copy_buffer_styled);
```
Comment (guedou) [inline src/helpers.c]:
```suggestion
    init_disassemble_info (info, stdout, (fprintf_ftype) copy_buffer, copy_buffer_styled);
```

**Comment by guedou** [comment]
Thanks! I made several modifications proposal. I think that we are ready for a release!

**Comment by CTM1** [comment]
Applied. Sorry that took so long, I've been quite busy
I'll be getting back to work on the library in the following weeks :)

**Comment by guedou** [comment]
No problem ! Do you want me to merge this first part?

**Comment by CTM1** [comment]
LGTM, fire away

**Comment by guedou** [comment]
Done!

**Pull request merged**